### ISSUE TITLE
False positive for `unsubscriptable-object`

### ISSUE THREAD

**Issue by JohnVillalovos** [PyCQA/pylint#6825]
### Bug description

With this file:

```
#!/usr/bin/python3 -ttu

from typing import Optional

class Reproducer:
    def __init__(self, name):
        self.http_username: Optional[str] = None
        self.set_values(name)

    def _get_values_from_helper(self) -> None:
        value = getattr(self, "http_username")
        if not isinstance(value, str):
            return

        helper = value[2:].strip()
        print(helper)


    def set_values(self, value: str) -> None:
        self.http_username = value
```

### Command used

```shell
pylint reproducer.py
```

### Pylint output

```shell
************* Module reproducer
reproducer.py:15:17: E1136: Value 'value' is unsubscriptable (unsubscriptable-object)

------------------------------------------------------------------
Your code has been rated at 6.15/10 (previous run: 3.08/10, +3.08)
```

### Expected behavior

Would expect it to realize `value` is a string.

### Pylint version

```shell
pylint 2.14.0
astroid 2.11.5
Python 3.10.4 (main, Mar 25 2022, 00:00:00) [GCC 11.2.1 20220127 (Red Hat 11.2.1-9)]
```

**Comment by jacobtylerwalls** [comment]
Thanks for the report. In one sense this is a duplicate of #1162, so I'll be closing it, but in another sense this is outside the scope of pylint:

`pylint` uses `astroid` as an inference engine to statically analyze your code to find what values variables actually take, not what they promise to take from type annotations. In this sample, only `None` is assigned to `http_username`. Of course, you have a setter, but despite the type annotation, it could take anything at runtime. If you want type-checking that performs sophisticated type-narrowing and incorporates your type annotations, `mypy` is your tool. 

That does mean, unfortunately, you will run into issues where you have to disable `no-member`, `unsubscriptable-object` in pylint from time to time.

We are open to improvements where we quiet pylint warnings when we detect ambiguity, and there are some interesting proposals in `astroid`, see https://github.com/PyCQA/astroid/pull/1189, but I don't see `astroid` implementing full-fledged type-narrowing and type annotation analysis like `mypy`--which is what would be needed to find "ambiguity" here in code that only ever sets `http_username` to `None`. I hope that makes sense.

**Comment by jacobtylerwalls** [comment]
Ah ha! Wait a darn minute! I feel like this is something that could be discussed in https://github.com/PyCQA/astroid/pull/1189, which introduces `Constraint`. If something previously inferable becomes `Uninferable` _only_ in virtue of a `Constraint`, we should be able to pass back another sentinel like `UninferableOnlyBecauseOfConstraintSoShutUpAboutNoMember`.

**Comment by jacobtylerwalls** [comment]
I'll go echo that comment on #1162 and https://github.com/PyCQA/astroid/pull/1189.

Despite my excitement I guess this is still a duplicate of either #1162 or #1498 or #4920.
